exercises/huffman/descr.md

```diff
@@ -0,0 +1,318 @@
+(immutable) priority queues. Four operations are provided. `empty` is the
+empty queue. `is_empty` tests whether a queue is empty. `insert x q` is the
+queue that contains the element `x` in the addition to the elements of the
+queue `q`. `extract` must be applied to a nonempty queue `q`. It returns a
+pair `(x, q')` where `x` is a *minimal* element of `q` (with respect to the
+user-specified total order on elements) and `q'` is the queue `q` deprived of
+`x`.
+
+**Question 2.**
+Define a module `Q` that provides priority queues whose
+elements are pairs of a tree and an integer frequency. Drawing
+an element of the queue must yield an element whose frequency
+is minimal.
+Then,
+Define the function `build_tree`,
+whose type is `alphabet -> tree`,
+which implements Huffman's algorithm,
+as described above.
+
+## Building an Encoding Dictionary
+
+As explained at the beginning of this exercise,
+a tree defines a correspondence between input characters
+and sequences of bits:
+indeed,
+a path from the root to a leaf corresponds to a bit string, where the bit
+`0` means "go down towards the left child" and the bit `1` means "go down
+towards the right child".
+
+In order to encode data in an efficient manner,
+we need a fast way of mapping each input character
+to its encoding, a bit string.
+In other words, we need an encoding dictionary:
+a map of every input character to its encoding.
+
+Such a map can be built in one traversal of the tree.
+
+**Question 3.**
+Define the function `build_encoding_dictionary`,
+whose type is `tree -> encoding_dictionary`.
+
+## Decoding Data
+
+In order to interpret a sequence of bits as a character,
+it suffices to view this sequence as a description of a
+path in the tree. By descending along this path, one reaches
+a leaf, which carries a character.
+
+**Question 4.** Write a function `find` of type
+`data -> int -> tree -> char * int`
+such that `find data i t`
+descends in the tree `t` along the path
+that is encoded at index `i` in the binary data `data`,
+until a leaf is reached,
+and returns the character carried by that leaf
+as well as an updated index into the binary data `data`.
+
+## Encoding and Decoding the Decoding Dictionary Itself
+
+Naturally, encoded data can be successfully decoded only
+if the decoding dictionary is available. Thus, when one
+wishes to store encoded data on disk or to transmit onto
+the network, one must store or transmit not only the
+encoded data, but also the decoding dictionary.
+
+Because the decoding dictionary is a tree, it must first
+be **serialized**, that is, transformed into binary data.
+
+At the other end, whoever wishes to decode the data must
+first **deserialize** the decoding dictionary, that is,
+read its binary representation and convert it back to a
+tree.
+
+**Question 5.**
+Complete the definitions of the functions `write`,
+whose type is `tree -> data`,
+and `read`,
+whose type is `data -> tree * int`.
```

Comment:
```suggestion
whose type is `data -> tree * int`. The returned integer is the index next to the end of the tree encoding in the input data.
```

Reply:
I would rather say "The integer result returned by `read` is the end index in the input data of the serialized dictionary (i.e., it is the index of whatever data follows the serialized dictionary)."